starknet_py/net/websockets/websocket_client.py

```diff
@@ -0,0 +1,369 @@
+            return False
+
+        params = {"subscription_id": subscription_id}
+        res = await self._send_message("starknet_unsubscribe", params)
+        unsubscription_result: bool = res["result"]
+
+        if unsubscription_result:
+            del self._subscriptions[subscription_id]
+
+        return unsubscription_result
+
+    async def _subscribe(
+        self,
+        handler: Callable[[Any], Any],
+        method: str,
+        params: Optional[Dict[str, Any]] = None,
+    ) -> str:
+        """ "
+        Creates a WebSocket stream which will fire events on a specific action.
+
+        :param handler: The function to call when a new notification is received.
+        :param method: The method to call to subscribe.
+        :param params: The parameters to pass to the method.
+
+        :return: The subscription ID.
+        """
+        response_message = await self._send_message(method, params)
+        subscription_id = response_message["result"]
+        self._subscriptions[subscription_id] = handler
+
+        return subscription_id
+
+    async def _listen(self):
+        """
+        Listens for incoming WebSocket messages.
+        """
+        if self.connection is None:
+            raise InvalidState("Connection is not established.")
+
+        async for message in self.connection:
+            self._handle_received_message(message)
+
+    async def _send_message(
+        self,
+        method: str,
+        params: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Sends a message to the WebSocket server.
+
+        :param method: The method to call.
+        :param params: The parameters to pass to the method.
+        """
+        if self.connection is None:
+            raise InvalidState("Connection is not established.")
+
+        message_id = self._message_id
+        self._message_id += 1
+
+        payload = {
+            "id": message_id,
+            "jsonrpc": "2.0",
+            "method": method,
+            "params": params if params else [],
+        }
+
+        future = asyncio.Future()
+        self._pending_responses[message_id] = future
+
+        await self.connection.send(json.dumps(payload))
+        response = await future
+
+        if "error" in response:
+            self._handle_error(response)
+
+        return response
+
+    def _handle_received_message(self, message: Union[str, bytes]):
+        """
+        Handles the received message.
+
+        :param message: The message received from the WebSocket server.
+        """
+        data = cast(Dict, json.loads(message))
+
+        # case when the message is a response to `subscribe_{method}`
+        if "id" in data and data["id"] in self._pending_responses:
+            future = self._pending_responses.pop(data["id"])
+            future.set_result(data)
+
+        # case when the message is a notification
+        elif "method" in data:
+            self._handle_notification(data)
```

Comment:
note: Both subscription responses and notifications are handled via `recv_streaming` (used in async iterator for the connection) because we cannot simultaneously use `recv` for immediate responses and maintain a continuous listener with `recv_streaming`. Consequently, all incoming messages are processed through the streaming interface.